spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,301 @@
+        throw new IllegalStateException("The option starting-snapshot-id " + startSnapshotId +
+            " is not an ancestor of the current snapshot");
+      }
+    }
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void setOffsetRange(Optional<Offset> start, Optional<Offset> end) {
+    table.refresh();
+
+    if (start.isPresent() && !StreamingOffset.START_OFFSET.equals(start.get())) {
+      this.startOffset = (StreamingOffset) start.get();
+      this.endOffset = (StreamingOffset) end.orElse(calculateEndOffset(startOffset));
+    } else {
+      // If starting offset is "START_OFFSET" (there's no snapshot in the last batch), or starting
+      // offset is not set, then we need to calculate the starting offset again.
+      this.startOffset = calculateStartingOffset();
+      this.endOffset = calculateEndOffset(startOffset);
+    }
+  }
+
+  @Override
+  public Offset getStartOffset() {
+    if (startOffset == null) {
+      throw new IllegalStateException("Start offset is not set");
+    }
+
+    return startOffset;
+  }
+
+  @Override
+  public Offset getEndOffset() {
+    if (endOffset == null) {
+      throw new IllegalStateException("End offset is not set");
+    }
+
+    return endOffset;
+  }
+
+  @Override
+  public Offset deserializeOffset(String json) {
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {
+    // Since all the data and metadata of Iceberg is as it is, nothing needs to commit when
+    // offset is processed, so no need to implement this method.
+  }
+
+  @Override
+  public void stop() {}
+
+  @Override
+  @SuppressWarnings("unchecked")
+  protected List<CombinedScanTask> tasks() {
+    if (startOffset.equals(endOffset)) {
+      LOG.info("Start offset {} equals to end offset {}, no data to process", startOffset, endOffset);
+      return Collections.emptyList();
+    }
+
+    Preconditions.checkState(pendingTasks != null,
+        "pendingTasks is null, which is unexpected as it will be set when calculating end offset");
+
+    if (Iterables.isEmpty(pendingTasks)) {
+      LOG.info("There's no task to process in this batch");
+      return Collections.emptyList();
+    }
+
+    IndexedTask last = Iterables.getLast(pendingTasks);
+    if (last.snapshotId() != endOffset.snapshotId() || last.index() != endOffset.index()) {
+      throw new IllegalStateException("The cached pendingTasks doesn't match the current end offset " + endOffset);
+    }
+
+    LOG.info("Processing data from {} exclusive to {} inclusive", startOffset, endOffset);
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(
+        CloseableIterable.transform(pendingTasks, IndexedTask::task), splitSize());
+    return Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize(), splitLookback(), splitOpenFileCost()));
+  }
+
+  private StreamingOffset calculateStartingOffset() {
```

Comment:
I would expect this to work as follows: if no `start_snapshot` is set, stream everything from the _current_ snapshot and then stream all subsequent incremental changes. If there is `start_snapshot`, stream only incremental changes starting from that snapshot.

Is there a reason to pick the first available snapshot and then stream incremental changes if no `start_snaphsot` is given?

Reply:
I don't have a specific preference. It is pretty like Kafka, either starting from the earliest offset or latest, I can change to either way or add a configuration to choose.

Reply:
There is one case where this might be important: users can do some preparations on existing tables before starting a stream. In that case, there will be multiple snapshots and some of them can have deletes.